File: lab6/project/lab-4-scala/src/main/scala/PatternMatching.scala
object PatternMatching {

  // Алгебраические типы данных
  sealed trait PaymentMethod
  case class CreditCard(number: String, expiry: String) extends PaymentMethod
  case class PayPal(email: String) extends PaymentMethod
  case class Crypto(wallet: String) extends PaymentMethod

  sealed trait OrderStatus
  case object Pending extends OrderStatus
  case object Processing extends OrderStatus
  case class Shipped(trackingNumber: String) extends OrderStatus
  case class Delivered(deliveryDate: String) extends OrderStatus
  case class Cancelled(reason: String) extends OrderStatus

  case class Order(id: Int, amount: Double, payment: PaymentMethod, status: OrderStatus)

  def processPayment(payment: PaymentMethod): String = payment match {
    case CreditCard(number, expiry) =>
      s"Обработка кредитной карты: ${number.takeRight(4)} (до $expiry)"
    case PayPal(email) =>
      s"Обработка PayPal: $email"
    case Crypto(wallet) =>
      s"Обработка криптовалюты: ${wallet.take(10)}..."
  }

  def canCancelOrder(status: OrderStatus): Boolean = status match {
    case Pending | Processing => true
    case Shipped(_) | Delivered(_) | Cancelled(_) => false
  }

  def updateOrderStatus(order: Order, newStatus: OrderStatus): Order = {
    order.copy(status = newStatus)
  }

  def demonstratePatternMatching(): Unit = {
    println("=== Pattern Matching ===")

    val orders = List(
      Order(1, 99.99, CreditCard("1234567812345678", "12/25"), Pending),
      Order(2, 149.99, PayPal("user@example.com"), Processing),
      Order(3, 199.99, Crypto("1A2b3C4d5E6f7G8h9I0j"), Shipped("TRACK123")),
      Order(4, 79.99, CreditCard("8765432187654321", "06/24"), Delivered("2024-01-15"))
    )

    // Обработка платежей
    orders.foreach { order =>
      val paymentInfo = processPayment(order.payment)
      val cancelable = if (canCancelOrder(order.status)) "можно отменить" else "нельзя отменить"
      println(s"Заказ ${order.id}: $paymentInfo - $cancelable")
    }

    // Деструктуризация в for-comprehension
    val pendingCreditCardOrders = for {
      order <- orders
      CreditCard(number, _) <- Some(order.payment) if canCancelOrder(order.status)
    } yield order

    println(s"Ожидающие заказы с кредитными картами: $pendingCreditCardOrders")

    // Сопоставление с образцом в map
    val statusDescriptions = orders.map {
      case Order(id, amount, _, Shipped(tracking)) =>
        s"Заказ $id отправлен, трекинг: $tracking"
      case Order(id, amount, _, Delivered(date)) =>
        s"Заказ $id доставлен $date"
      case Order(id, amount, _, status) =>
        s"Заказ $id в статусе: $status"
    }

    statusDescriptions.foreach(println)
  }

  def main(args: Array[String]): Unit = {
    demonstratePatternMatching()
  }
}

File: lab6/project/lab-5-rust/src/pattern_matching.rs
// Алгебраические типы данных
#[derive(Debug)]
enum PaymentMethod {
    CreditCard { number: String, expiry: String },
    PayPal { email: String },
    Crypto { wallet: String },
}

#[derive(Debug)]
enum OrderStatus {
    Pending,
    Processing,
    Shipped(String), // tracking number
    Delivered(String), // delivery date
    Cancelled { reason: String },
}

#[derive(Debug)]
struct Order {
    #[allow(dead_code)]
    id: u32,
    #[allow(dead_code)]
    amount: f64,
    payment: PaymentMethod,
    status: OrderStatus,
}

impl Order {
    fn new(id: u32, amount: f64, payment: PaymentMethod, status: OrderStatus) -> Self {
        Order { id, amount, payment, status }
    }
}

fn process_payment(payment: &PaymentMethod) -> String {
    match payment {
        PaymentMethod::CreditCard { number, expiry } => {
            let last_four = number.chars().rev().take(4).collect::<String>();
            format!("Обработка кредитной карты: ****{} (до {})", last_four, expiry)
        }
        PaymentMethod::PayPal { email } => {
            format!("Обработка PayPal: {}", email)
        }
        PaymentMethod::Crypto { wallet } => {
            let shortened = wallet.chars().take(10).collect::<String>();
            format!("Обработка криптовалюты: {}...", shortened)
        }
    }
}

fn can_cancel_order(status: &OrderStatus) -> bool {
    match status {
        OrderStatus::Pending | OrderStatus::Processing => true,
        OrderStatus::Shipped(_) | OrderStatus::Delivered(_) | OrderStatus::Cancelled { .. } => false,
    }
}

fn update_order_status(order: Order, new_status: OrderStatus) -> Order {
    Order { status: new_status, ..order }
}

fn demonstrate_pattern_matching() {
    println!("\n=== Pattern Matching ===");
    
    let orders = vec![
        Order::new(
            1, 
            99.99, 
            PaymentMethod::CreditCard { 
                number: "1234567812345678".to_string(), 
                expiry: "12/25".to_string() 
            }, 
            OrderStatus::Pending
        ),
        Order::new(
            2, 
            149.99, 
            PaymentMethod::PayPal { 
                email: "user@example.com".to_string() 
            }, 
            OrderStatus::Processing
        ),
        Order::new(
            3, 
            199.99, 
            PaymentMethod::Crypto { 
                wallet: "1A2b3C4d5E6f7G8h9I0j".to_string() 
            }, 
            OrderStatus::Shipped("TRACK123".to_string())
        ),
    ];
    
    // Обработка заказов с pattern matching
    for order in &orders {
        let payment_info = process_payment(&order.payment);
        let cancelable = if can_cancel_order(&order.status) { 
            "можно отменить" 
        } else { 
            "нельзя отменить" 
        };
        println!("Заказ {}: {} - {}", order.id, payment_info, cancelable);
    }
    
    // Деструктуризация в if let
    for order in orders {
        if let OrderStatus::Shipped(tracking) = &order.status {
            println!("Заказ {} отправлен, трекинг: {}", order.id, tracking);
        }
        
        // Сопоставление с несколькими паттернами
        match order.status {
            OrderStatus::Delivered(date) => {
                println!("Заказ {} доставлен {}", order.id, date);
            }
            OrderStatus::Cancelled { reason } => {
                println!("Заказ {} отменен: {}", order.id, reason);
            }
            _ => {} // Игнорируем другие статусы
        }
    }
}

fn main() {
    
    demonstrate_pattern_matching();
}